### DISCUSSION TITLE
Does anyone knows if something similar to this app has a Linux Desktop alternative?

### DISCUSSION THREAD

**Discussion by iHarryPotter178** [mateusz-bak/openreads#770] [General]
It's not exactly a feature request, but I like this app and what it does, and want to do something similar on my Laptop to keep track of things. So I'm looking for a Desktop App which is available for Linux? Any Help would be appreciated.

**Comment by exstntlmsnthrp** [comment]
I've searched for a very long time and have never found anything like this for Linux.

**Comment by iHarryPotter178** [comment]
> I've searched for a very long time and have never found anything like this for Linux.

Then, Waydroid is the solution, or do you use something else for this purpose?

**Comment by exstntlmsnthrp** [comment]
> > I've searched for a very long time and have never found anything like this for Linux.
> 
> Then, Waydroid is the solution, or do you use something else for this purpose?

I do not use it on Linux, only on my Android device. Besides, it's very convenient to use the barcode scanner when adding books.

**Comment by lucastucious** [comment]
I still use [GCStar](https://gitlab.com/GCstar/GCstar), that is not really the same but with the same purpose.

**Comment by mateusz-bak** [comment]
Openreads is written in Flutter which supports Linux Desktop. If there are people that would use it I can try to add the platform.

**Comment by iHarryPotter178** [comment]
That will be great.. Please do add it.. 

> Openreads is written in Flutter which supports Linux Desktop. If there are people that would use it I can try to add the platform.

**Comment by lee-b** [comment]
> Openreads is written in Flutter which supports Linux Desktop. If there are people that would use it I can try to add the platform.

This would be excellent, I think.  Or perhaps even a server mode that serves the UI over the web, if that's possible with Flutter. I think most people would prefer to edit book details on a desktop machine, and only browse the collection or update reading status on mobile, or quickly add books to their reading list while they're talking to people, but still to fill in the details later on desktop.

**Comment by armand-leguillou** [comment]
Hi! I plan to no longer have a mobile phone in the future, but Openreads is essential! If a desktop version is possible and if the developpers have enough time, it could be great! Thanks for your work and for the app.

**Comment by RomanKovtyukh** [comment]
Hi @mateusz-bak, I've tried running it on linux.

### System
 
* `$ flutter --version`
```
Flutter 3.29.0 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 35c388afb5 (4 months ago) • 2025-02-10 12:48:41 -0800
Engine • revision f73bfc4522
Tools • Dart 3.7.0 • DevTools 2.42.2
```
* `$ flutter doctor` 
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.29.0, on Arch Linux 6.14.7-arch2-1, locale en_US.UTF-8)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
...
[✓] Chrome - develop for the web
[✓] Linux toolchain - develop for Linux desktop
[!] Android Studio (not installed)
[✓] Connected device (2 available)
[✓] Network resources
! Doctor found issues in 2 categories.
```
* `flutter devices`
```
Found 2 connected devices:
  Linux (desktop) • linux  • linux-x64      • Arch Linux 6.14.7-arch2-1
  Chrome (web)    • chrome • web-javascript • Google Chrome 136.0.7103.113
```

### Running

```
flutter pub get
flutter create .
flutter -d linux run
```

### Observations

1. At startup it fails to create database
```
Launching lib/main.dart on Linux in debug mode...
Building Linux application...                                           
✓ Built build/linux/x64/debug/bundle/openreads

(software.mdev.openreads:52619): Atk-CRITICAL **: 16:10:46.343: atk_socket_embed: assertion 'plug_id != NULL' failed
flutter: [🌎 Easy Localization] [DEBUG] Localization initialized
flutter: [🌎 Easy Localization] [DEBUG] Start
[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: Bad state: databaseFactory not initialized
databaseFactory is only initialized when using sqflite. When using `sqflite_common_ffi`
You must call `databaseFactory = databaseFactoryFfi;` before using global openDatabase API

#0      databaseFactory.<anonymous closure> (package:sqflite_common/src/sqflite_database_factory.dart:29:7)
#1      databaseFactory (package:sqflite_common/src/sqflite_database_factory.dart:33:6)
#2      openDatabase (package:sqflite_common/sqflite.dart:83:10)
#3      DatabaseProvider.createDatabase (package:openreads/database/database_provider.dart:29:18)
<asynchronous suspension>
#4      DatabaseController.getAllBooks (package:openreads/database/database_controler.dart:30:16)
<asynchronous suspension>
#5      BookCubit._migrateCoversFromDatabaseToStorage (package:openreads/logic/cubit/book_cubit.dart:274:27)
<asynchronous suspension>
#6      BookCubit._initLoad (package:openreads/logic/cubit/book_cubit.dart:71:7)
<asynchronous suspension>

flutter: [🌎 Easy Localization] [DEBUG] Init state
flutter: [🌎 Easy Localization] [DEBUG] Build
flutter: [🌎 Easy Localization] [DEBUG] Init Localization Delegate
flutter: [🌎 Easy Localization] [DEBUG] Init provider
flutter: [🌎 Easy Localization] [DEBUG] Load Localization Delegate
flutter: [🌎 Easy Localization] [DEBUG] Load asset from assets/translations
flutter: [🌎 Easy Localization] [DEBUG] Load asset from assets/translations
flutter: dynamic_color: Accent color detected.
flutter: [🌎 Easy Localization] [WARNING] Unable to load asset: "assets/translations/en.json".
The asset does not exist or has empty data.
flutter: [🌎 Easy Localization] [DEBUG] Load asset from assets/translations
Syncing files to device Linux...                                   186ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

A Dart VM Service on Linux is available at: http://127.0.0.1:38979...
The Flutter DevTools debugger and profiler on Linux is available at: http://127.0.0.1:9100...
```
1. On the first run it looked like first run on an android phone: four introduction screens one of which suggesting to restore/import database.
2. [ISSUE] None of those import buttons worked (just nothing happened). Maybe because of the database factory failures above
3. On the second run it didn't show the introduction screens; so it was able to write something on the disk, but I couldn't find where to delete it and go back to those screens
4. "Finished" screen hang on loading something (perhaps data from database)
5. settings menu leads to the settings screen
6. on the settings screen:
  - worked well:
    - join the community -- opened the link in default system browser
    - help with translation -- same
  - didn't work
    - support the development: can't open the menu with "become github supporter" and "buy me a coffee"
    - send feedback
    - rate the application
    - deleted books: tried working with database factory and failed, but switching to deleted books screen and back worked well.

The DatabaseFactory issue seems to be the first thing to look at.

PS: don't like the flutter installation method so far; to me downloading an archive and extracting it in a random place is a degradation in package management.

**Comment by RomanKovtyukh** [comment]
Created a PR with the sqlite configuration for linux (and windows) desktops. It's still failing to add new books (+ button doesn't work), but the "Finished" screen is not hanging on loading any more and no errors at the start.

**Comment by RomanKovtyukh** [comment]
Ok, the reason why the "Add book" button wasn't working is [here](https://github.com/mateusz-bak/openreads/blob/master/lib/ui/books_screen/books_screen.dart#L809-L835): the code is written only for ios and android. Making android code a default branch and leaving ios under `if` worked: it was able to show the bottom menu and even switch to the "Search in open library" page.

Search page then worked fine to find a book by ISBN and to show it to me, but then the app crashed after I pressed "Add book". Switched to another view and crashed to be precise.

Failed [here](https://github.com/mateusz-bak/openreads/blob/master/lib/core/helpers/helpers.dart#L17)
```
Exception has occurred.
MissingPluginException (MissingPluginException(No implementation found for method blurHashEncode on channel blurhash))
```

Adding books manually works fine though. At least with the fields I quickly filled. The books didn't load at the next run.

**Comment by lucastucious** [comment]
Maybe books aren't shown because the database creation failed on build ?

**Comment by RomanKovtyukh** [comment]
I think I must've misinterpreted some behaviors, but now I can't reproduce the books not loading on next run. Adding manually works well, adding through OpenLibrary got fixed by using `blurhash_fft` for building a cover hash for the database.

**Comment by RomanKovtyukh** [comment]
Updated https://github.com/mateusz-bak/openreads/pull/678 with those changes.

**Comment by RomanKovtyukh** [comment]
Tried syncing with android DB:
1. couldn't get the [`Books.db`](https://github.com/mateusz-bak/openreads/blob/68a0f21db9262d48f65d6e51c748cf14d591cbc8/lib/logic/bloc/migration_v1_to_v2_bloc/migration_v1_to_v2_bloc.dart#L80) on my phone. Seems like it's [not available to other applications for security reasons](https://stackoverflow.com/a/51444404/2114651). So if we want [syncthing](https://github.com/mateusz-bak/openreads/issues/462) the DB itself we'll probably have to store it somewhere else. Maybe the directory used for the export and backups can be a good place.
2. couldn't use neither "import" nor "restore from backup" because it doesn't open any dialog to choose file

**Comment by RomanKovtyukh** [comment]
Import or restore didn't work because the [`AlertDialog.adaptive`](https://github.com/mateusz-bak/openreads/blob/68a0f21db9262d48f65d6e51c748cf14d591cbc8/lib/ui/settings_screen/settings_backup_screen.dart#L373) didn't show anything on Linux. Replacing with `AlertDialog` fixed the issue for Linux, but this means many more `if not linux` allover the app :(

Then after pressing "Yes" on the alert dialog we have one more `if android {} else if ios {}` [in the restore code](https://github.com/mateusz-bak/openreads/blob/68a0f21db9262d48f65d6e51c748cf14d591cbc8/lib/ui/settings_screen/settings_backup_screen.dart#L144-L153). And there are [around 10 potential places like this](https://github.com/search?q=repo%3Amateusz-bak%2Fopenreads%20Platform.isAndroid&type=code).

So for now it doesn't look like "just adding linux/other desktop as a build target". At the same time nothing really blocking so far.

**Comment by RomanKovtyukh** [comment]
Ok, made it work: standard file picking tool used in ios worked on linux well and then restoration worked. There were couple of issues with restoration, but they seemed not platform-specific:
* it said that backup file is not compatible first, but then worked well on second attempt
* it wasn't able to read `25.jpg` and it's really not in the archive; problem with building backup maybe

Need to fix the `AlertDialog.adaptive` properly before updating the PR.

**Comment by armand-leguillou** [comment]
Do you think it will be possible to create a real linux version available on Flathub etc? Anyways, well done and thank for your work!

**Comment by RomanKovtyukh** [comment]
Yes, it doesn't look unrealistic. Most of the issues I've faced so far were because of the code written without Linux support in mind. The ones which came from a third party or from the Flutter itself were quite solvable.

I think in worst case we'll have some functions not working, but what I have now is already quite usable. Need to fix the `AlertDialog.adaptive` properly first though. And we'll need Mateusz to review it before we can think of packaging.

**Comment by RomanKovtyukh** [comment]
Ok, so the last version of the PR has all `if (Platform.isAndroid/IOS)` adjusted to work on linux or return an understandable error to users.

Can anyone try building it on their machine? I'll check later if I can build a flatpak or something.

**Comment by RomanKovtyukh** [comment]
Ok, I'll stop at this point until someone gives it a review. Just to see what will be the feedback before moving on. One of the next steps would be to make mobile version store the database somewhere else to allow syncthing to do it's business (https://github.com/mateusz-bak/openreads/issues/462).

**Comment by armand-leguillou** [comment]
hi! i can test it on my pc!
but i don't no how to do it: what do i have to download? (i am on fedora 42)

**Comment by GLLM** [comment]
smae here. I would be OK to test. Fedora 42 KDE here.
Thanks

**Comment by RomanKovtyukh** [comment]
I couldn't find artifacts from the PR builds: https://github.com/mateusz-bak/openreads/actions/runs/15690018649/job/44202665471?pr=678. So I'll need to figure out how to build it on github. Don't really want to share any binaries from my machine here. Maybe will build it in my fork.

**Comment by armand-leguillou** [comment]
ok, thanks for your implication ! keep us up informed!

**Comment by ironhak** [comment]
Just dropping by to show interest for a Linux version :)

**Comment by armand-leguillou** [comment]
any update?

**Reply by RomanKovtyukh** [reply]
Hi. Sorry, need to add automatic linux build to the PR. The code itself is more or less running.
